### ISSUE TITLE
Rule syntax error causing panic when using functions that should return the errors

### ISSUE THREAD

**Issue by newm4n** [hyperjumptech/grule-rule-engine#214]
### Discussed in https://github.com/hyperjumptech/grule-rule-engine/discussions/213

<div type='discussions-op-text'>

<sup>Originally posted by **ibtrigger** July 14, 2021</sup>
I'm in the process of building a web interface for our team to add/modify/delete rule files on a remote server. The interface allows for a simple "syntax check" for a rule prior to testing the rule. I am using the "BuildRuleFromResource" function from the rule builder which is supposed to return any errors that occur. It is not consistent on when errors are returned as expected  or when a low level function "panics". When the panic occurs the errors are caught as expected and logged out from the running process, but the errors are not returned by the function and I am unable to send the actual syntax error messages back to the user on the web interface.

The panic shows interface errors. I have created a very minimal example of code that demonstrates the panic errors. Here is the sample code:

```
package main

import (
  "fmt"
  "github.com/hyperjumptech/grule-rule-engine/ast"
  "github.com/hyperjumptech/grule-rule-engine/builder"
  "github.com/hyperjumptech/grule-rule-engine/pkg"
)

func main() {
  GRL :=
`rule ExampleRuleName "One line rule description" salience 0  {
  when
    <FACT.SomeField == "true"
  then
    FACT.WriteMessage*( "message to write" );
}`

  lib := ast.NewKnowledgeLibrary()
  ruleBuilder := builder.NewRuleBuilder( lib )
  err := ruleBuilder.BuildRuleFromResource( "SimpleTest", "1.0", pkg.NewBytesResource( []byte( GRL ) ) )
  if err == nil {
    fmt.Println( "Rule built successfully" )
  } else {
    fmt.Printf( "Error building rule: %v\n", err )
  }
}
```

You can see that I intentionally inserted a "<" in the when clause of the rule to force a syntax error. When I run the test I get the following:

```
ERRO[0000] GRL error, after '["One line rule description" salience 0 { when]' and then unexpected '<'  lib=grule struct=GruleParserV3Listener
panic: interface conversion: *ast.Expression is not ast.MemberVariableReceiver: missing method AcceptMemberVariable

goroutine 1 [running]:
github.com/hyperjumptech/grule-rule-engine/antlr.(*GruleV3ParserListener).ExitMemberVariable(0xc0001ce380, 0xc00028a1a0)
	/Users/XXXXXXX/go/pkg/mod/github.com/hyperjumptech/grule-rule-engine@v1.9.1/antlr/GruleParserV3Listener.go:531 +0x53
github.com/hyperjumptech/grule-rule-engine/antlr/parser/grulev3.(*MemberVariableContext).ExitRule(0xc00028a1a0, 0x158bf80, 0xc0001ce380)
	/Users/XXXXXXX/go/pkg/mod/github.com/hyperjumptech/grule-rule-engine@v1.9.1/antlr/parser/grulev3/grulev3_parser.go:3310 +0x73
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).ExitRule(0x1928788, 0x158bf80, 0xc0001ce380, 0x1f5d260, 0xc00028a1a0)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:252 +0x77
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158e980, 0xc00028a1a0)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:233 +0x136
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158ee60, 0xc00028a160)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:231 +0xc3
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158e680, 0xc00021d7e0)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:231 +0xc3
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158e6e0, 0xc00021c520)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:231 +0xc3
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158e6e0, 0xc00028b3e0)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:231 +0xc3
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158eec0, 0xc00021c380)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:231 +0xc3
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158ebc0, 0xc0001ef460)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:231 +0xc3
github.com/antlr/antlr4/runtime/Go/antlr.(*ParseTreeWalker).Walk(0x1928788, 0x158bf80, 0xc0001ce380, 0x158e800, 0xc0001d3320)
	/Users/XXXXXXX/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200801005519-2ba38605b949/runtime/Go/antlr/tree.go:231 +0xc3
github.com/hyperjumptech/grule-rule-engine/builder.(*RuleBuilder).BuildRuleFromResource(0xc0000f1e48, 0x14ef018, 0xa, 0x14ed247, 0x3, 0x1586f80, 0xc0001d3280, 0x7, 0x0)
	/Users/XXXXXXX/go/pkg/mod/github.com/hyperjumptech/grule-rule-engine@v1.9.1/builder/RuleBuilder.go:128 +0x53c
main.main()
	/Users/XXXXXXX/go/src/test/rules.go:21 +0x17a
```

It makes a difference where the syntax errors occur on whether the error messages will be successfully returned as expected by the "BuildRuleFromResource"  function. Syntax errors in some locations within the rule return the errors and some do not. 

Here are the associated panic messages I have encountered while testing that indicate the missing interface functions:

```
panic: interface conversion: *ast.Expression is not ast.MemberVariableReceiver: missing method AcceptMemberVariable
panic: interface conversion: *ast.Salience is not ast.MemberVariableReceiver: missing method AcceptMemberVariable
panic: interface conversion: *ast.ThenExpressionList is not ast.MemberVariableReceiver: missing method AcceptMemberVariable
```

It appears that some low level structures do not provide all the required functions that they should have to fulfill interface requirements. 

FYI: Running GO 1.14 on MacOS
</div>

**Comment by newm4n** [comment]
@niallnsec I've checked parser hardening branch you made and its much better than mine. Looking forward for your PR.

**Comment by newm4n** [comment]
I decided to squash merged parser-hardening branch directly to master my self. Its been tested and I've reviewed it.
